components/script/window_named_properties.rs

```diff
@@ -0,0 +1,205 @@
+            nativeCall: None,
+            hasInstance: None,
+            objectClassIs: None,
+            className: Some(class_name),
+            fun_toString: None,
+            boxedValue_unbox: None,
+            defaultValue: None,
+            trace: None,
+            finalize: None,
+            objectMoved: None,
+            isCallable: None,
+            isConstructor: None,
+        };
+
+        #[allow(unsafe_code)]
+        unsafe {
+            SyncWrapper(CreateProxyHandler(&traps, ptr::null()))
+        }
+    };
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn get_own_property_descriptor(
+    cx: *mut JSContext,
+    proxy: HandleObject,
+    id: HandleId,
+    desc: MutableHandle<PropertyDescriptor>,
+) -> bool {
+    let cx = SafeJSContext::from_ptr(cx);
+    if !RUST_JSID_IS_STRING(id) {
+        // Nothing to do if we're resolving a non-string property.
+        return true;
+    }
+
+    let mut found = false;
+    if !has_property_on_prototype(
+        *cx,
+        RustHandle::from_raw(proxy),
+        RustHandle::from_raw(id),
+        &mut found,
+    ) {
+        return false;
+    }
+    if found {
+        return true;
+    }
+
+    let s = jsstr_to_string(*cx, RUST_JSID_TO_STRING(id));
+    if s.is_empty() {
+        return true;
+    }
+
+    let window = Root::downcast::<Window>(GlobalScope::from_object(proxy.get()))
+        .expect("global is not a window");
+    if let Some(obj) = window.NamedGetter(cx, s.into()) {
+        rooted!(in(*cx) let mut val = UndefinedValue());
+        obj.to_jsval(*cx, val.handle_mut());
+        fill_property_descriptor(RustMutableHandle::from_raw(desc), proxy.get(), val.get(), 0);
+    }
+    return true;
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn define_property(
+    cx: *mut JSContext,
+    _proxy: HandleObject,
+    _id: HandleId,
+    _desc: Handle<PropertyDescriptor>,
+    _result: *mut ObjectOpResult,
+) -> bool {
+    throw_type_error(
+        cx,
+        "Not allowed to define a property on the named properties object.",
+    );
+    false
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn delete(
+    _cx: *mut JSContext,
+    _proxy: HandleObject,
+    _id: HandleId,
+    result: *mut ObjectOpResult,
+) -> bool {
+    (*result).code_ = JSErrNum::JSMSG_CANT_DELETE_WINDOW_NAMED_PROPERTY as usize;
+    true
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn prevent_extensions(
+    _cx: *mut JSContext,
+    _proxy: HandleObject,
+    result: *mut ObjectOpResult,
+) -> bool {
+    (*result).code_ = JSErrNum::JSMSG_CANT_PREVENT_EXTENSIONS as usize;
+    true
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn is_extensible(
+    _cx: *mut JSContext,
+    _proxy: HandleObject,
+    extensible: *mut bool,
+) -> bool {
+    *extensible = true;
+    true
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn class_name(_cx: *mut JSContext, _proxy: HandleObject) -> *const i8 {
+    &b"WindowProperties" as *const _ as *const i8
```

Comment:
This string needs a `\0` final byte.

Reply:
Great catch! That changed the test results from `CRASH` to `TIMEOUT`. I'd call that progress.

Thanks for your effort looking through the changes!